src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -300,161 +146,172 @@ public void UpdateTrackedDocument(Uri uri, SourceText newSourceText)
+            if (documents.Any())
             {
-                return null;
-            }
+                var document = documents.FindDocumentInProjectContext(textDocumentIdentifier);
 
-            // Filter the matching documents by project context.
-            var documentInProjectContext = findDocumentResult.FindDocumentInProjectContext(textDocumentIdentifier);
-            return documentInProjectContext;
+                // Record metadata on how we got this document.
+                var workspaceKind = document.Project.Solution.Workspace.Kind;
+                _requestTelemetryLogger.UpdateFindDocumentTelemetryData(success: true, workspaceKind);
+                _requestTelemetryLogger.UpdateUsedForkedSolutionCounter(isForked);
+                _logger.TraceInformation($"{document.FilePath} found in workspace {workspaceKind}");
+
+                return document;
+            }
         }
-    }
 
-    #endregion
+        // We didn't find the document in any workspace, record a telemetry notification that we did not find it.
+        var searchedWorkspaceKinds = string.Join(";", lspSolutions.SelectAsArray(lspSolution => lspSolution.Solution.Workspace.Kind));
+        _logger.TraceError($"Could not find '{uri}'.  Searched {searchedWorkspaceKinds}");
+        _requestTelemetryLogger.UpdateFindDocumentTelemetryData(success: false, workspaceKind: null);
+
+        // Add the document to our loose files workspace if its open.
+        var miscDocument = _trackedDocuments.ContainsKey(uri) ? _lspMiscellaneousFilesWorkspace?.AddMiscellaneousDocument(uri, _trackedDocuments[uri]) : null;
+        return miscDocument;
+    }
 
     /// <summary>
-    /// Helper to clear out the LSP incremental solution for all registered workspaces.
-    /// Should be called under <see cref="_gate"/>
+    /// Gets the LSP view of all the registered workspaces' current solutions.
     /// </summary>
-    private void ResetIncrementalLspSolutions_CalledUnderLock()
+    private ImmutableArray<(Solution Solution, bool IsForked)> GetLspSolutions()
     {
-        Contract.ThrowIfFalse(Monitor.IsEntered(_gate));
-
-        var workspaces = _workspaceToLspSolution.Keys.ToImmutableArray();
-        foreach (var workspace in workspaces)
+        // Ensure that the loose files workspace is searched last.
+        var registeredWorkspaces = _lspWorkspaceRegistrationService.GetAllRegistrations();
+        registeredWorkspaces = registeredWorkspaces
+            .Where(workspace => workspace is not LspMiscellaneousFilesWorkspace)
+            .Concat(registeredWorkspaces.Where(workspace => workspace is LspMiscellaneousFilesWorkspace)).ToImmutableArray();
+
+        using var _ = ArrayBuilder<(Solution, bool)>.GetInstance(out var solutions);
+        foreach (var workspace in registeredWorkspaces)
         {
-            _workspaceToLspSolution[workspace] = null;
+            var workspaceCurrentSolution = workspace.CurrentSolution;
```

Comment:
doc the idea on this being safe/correct even though workspace.CurrentSollution being something that could be mutating while this is being called.